### PR TITLE
Refactor/fullrun

### PR DESCRIPTION
**Pull request by orianac**

### CODE REVIEW

File: cmip6_downscaling/methods/common/utils.py [hunk 1]
```diff
@@ -97,22 +97,54 @@ def calc_auspicious_chunks_dict(
...
+        The dataset of monthly spatial anomalies resulting from taking the difference between
+        the fine scale obs and the interpolated obs. With dimensions ('month', 'lat', 'lon')
+
+    Returns
+    -------
+    reconstructed : xr.Dataset
+        Finescale dataset with spatial heterogeneity added back in
+    """
+    reconstructed = ds.groupby('time.month') + spatial_anomaly
+    return reconstructed
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/utils.py:150]: Can we move this to bcsd?

File: cmip6_downscaling/methods/bcsd/flow.py [hunk 2]
```diff
@@ -69,27 +73,31 @@
...
     coarse_obs_full_time_path = rechunk(coarse_obs_path, chunking_pattern='full_time')
-    experiment_train_full_time_path = rechunk(
-        experiment_train_path, chunking_pattern=coarse_obs_full_time_path
-    )
+    experiment_train_full_time_path = rechunk(experiment_train_path, chunking_pattern='full_time')
     experiment_predict_full_time_path = rechunk(
-        experiment_predict_path, chunking_pattern=coarse_obs_full_time_path
+        experiment_predict_path,
+        chunking_pattern='full_time',
+        chunking_template_file=coarse_obs_full_time_path,
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd/flow.py:80]:
```suggestion
        pattern='full_time',
        template=coarse_obs_full_time_path,
```

Comment (jhamman) [reply cmip6_downscaling/methods/bcsd/flow.py:80]: repeat this throughout

File: cmip6_downscaling/methods/common/tasks.py [hunk 3]
```diff
@@ -144,16 +144,26 @@ def get_experiment(run_parameters: RunParameters, time_subset: str) -> UPath:
 
 
 @task(log_stdout=True)
-def rechunk(path: UPath, chunking_pattern: Union[str, UPath] = None, max_mem: str = "2GB") -> UPath:
+def rechunk(
+    path: UPath,
+    chunking_pattern: str = None,
+    chunking_template_file: UPath = None,
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/tasks.py:150]:
```suggestion
    pattern: str = None,
    template: UPath = None,
```

File: cmip6_downscaling/methods/common/tasks.py [hunk 4]
```diff
@@ -144,16 +144,26 @@ def get_experiment(run_parameters: RunParameters, time_subset: str) -> UPath:
...
     ----------
     path : UPath
         path to zarr store
-    chunking_pattern : str or UPath
-        The pattern of chunking you want to use.
+    chunking_pattern : str
+        The pattern of chunking you want to use. If used together with `chunking_template_file` it will override the template
+        to ensure that the final dataset truly follows that `full_space` or `full_time` spec. This matters when you are passing
+        a template that is either a shorter time length or a template that is a coarser grid (and thus a shorter lat/lon chunksize)
+    chunking_template_file : UPath
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/tasks.py:164]:
```suggestion
    pattern : str
        The pattern of chunking you want to use. If used together with `template` it will override the template
        to ensure that the final dataset truly follows that `full_space` or `full_time` spec. This matters when you are passing
        a template that is either a shorter time length or a template that is a coarser grid (and thus a shorter lat/lon chunksize)
    template : UPath
```

File: cmip6_downscaling/methods/common/tasks.py [hunk 5]
```diff
@@ -162,21 +172,20 @@ def rechunk(path: UPath, chunking_pattern: Union[str, UPath] = None, max_mem: st
...
-
-    if isinstance(chunking_pattern, str):
-        pattern_string = chunking_pattern
-    elif isinstance(chunking_pattern, UPath):
-        # when you pass a dataset the chunking will match the chunking of that dataset
-        # so we'll call it "matched". this is ambiguous - like it could be chunked to match
-        # any number of patterns but it is sufficient for this implementation.
+    # print('rechunking dataset at {}'.format(path))
+    # if both defined then you'll take the spatial part of template and override one dimension with the specified pattern
+    if chunking_template_file is not None:
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/tasks.py:177]:
```suggestion
    if template is not None:
```

Comment (jhamman) [reply cmip6_downscaling/methods/common/tasks.py:177]: apply elsewhere in this function too

File: cmip6_downscaling/methods/common/utils.py (hunk 1 above)
Comment (orianac) [reply cmip6_downscaling/methods/common/utils.py:150]: moved to `cmip6_downscaling.methods.bcsd.utils`

File: cmip6_downscaling/methods/bcsd/flow.py (hunk 2 above)
Comment (orianac) [reply cmip6_downscaling/methods/bcsd/flow.py:80]: done

File: cmip6_downscaling/methods/common/tasks.py (hunk 3 above)
Comment (orianac) [reply cmip6_downscaling/methods/common/tasks.py:150]: done

File: cmip6_downscaling/methods/common/tasks.py [hunk 6]
```diff
@@ -471,20 +485,13 @@ def pyramid(ds_path: UPath, levels: int = 2, other_chunks: dict = None) -> UPath
...
-    dta = _pyramid_postprocess(dta, levels, other_chunks=other_chunks, ds_name=ds_name)
+        dta = _pyramid_postprocess(dta, levels, other_chunks=other_chunks, ds_name=ds_name)
 
-    # write to target
-    dta.to_zarr(target, mode='w')
-    return target
-
-
-@task
-def run_analyses(ds_path: UPath, run_parameters: RunParameters) -> UPath:
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/tasks.py:487]: yikes! This was not meant to happen. @andersy005 is going to revert these changes.
